service/history/handler_test.go

```diff
@@ -148,3 +150,143 @@ func (s *handlerSuite) testRespondCrossClusterTaskCompleted(
+			},
+			expectation: func(scope *mocks.Scope) {
+				scope.Mock.On("IncCounter", metrics.CadenceErrDomainNotActiveCounter).Once()
+			},
+		},
+		"workflow is already started err": {
+			input: &types.WorkflowExecutionAlreadyStartedError{
+				Message: "bad already started",
+			},
+			expectation: func(scope *mocks.Scope) {
+				scope.Mock.On("IncCounter", metrics.CadenceErrExecutionAlreadyStartedCounter).Once()
+			},
+		},
+		"does not exist": {
+			input: &types.EntityNotExistsError{
+				Message: "the workflow doesn't exist",
+			},
+			expectation: func(scope *mocks.Scope) {
+				scope.Mock.On("IncCounter", metrics.CadenceErrEntityNotExistsCounter).Once()
+			},
+		},
+		"already completed": {
+			input: &types.WorkflowExecutionAlreadyCompletedError{
+				Message: "the workflow is done",
+			},
+			expectation: func(scope *mocks.Scope) {
+				scope.Mock.On("IncCounter", metrics.CadenceErrWorkflowExecutionAlreadyCompletedCounter).Once()
+			},
+		},
+		"Cancellation already requested": {
+			input: &types.CancellationAlreadyRequestedError{
+				Message: "the workflow is cancelled already",
+			},
+			expectation: func(scope *mocks.Scope) {
+				scope.Mock.On("IncCounter", metrics.CadenceErrCancellationAlreadyRequestedCounter).Once()
+			},
+		},
+		"rate-limits": {
+			input: &types.LimitExceededError{
+				Message: "limits exceeded",
+			},
+			expectation: func(scope *mocks.Scope) {
+				scope.Mock.On("IncCounter", metrics.CadenceErrLimitExceededCounter).Once()
+			},
+		},
+		"retry tasks": {
+			input: &types.RetryTaskV2Error{
+				Message: "limits exceeded",
+			},
+			expectation: func(scope *mocks.Scope) {
+				scope.Mock.On("IncCounter", metrics.CadenceErrRetryTaskCounter).Once()
+			},
+		},
+		"service busy error": {
+			input: &types.ServiceBusyError{
+				Message: "limits exceeded - service is busy",
+			},
+			expectation: func(scope *mocks.Scope) {
+				scope.Mock.On("IncCounter", metrics.CadenceErrServiceBusyCounter).Once()
+			},
+		},
+		"deadline exceeded": {
+			input: yarpcerrors.DeadlineExceededErrorf("some deadline exceeded err"),
+			expectation: func(scope *mocks.Scope) {
+				scope.Mock.On("IncCounter", metrics.CadenceErrContextTimeoutCounter).Once()
+				scope.Mock.On("IncCounter", metrics.CadenceFailures).Once()
+			},
+		},
+		"internal error": {
+			input: types.InternalServiceError{Message: "internal error"},
+			expectation: func(scope *mocks.Scope) {
+				scope.Mock.On("IncCounter", metrics.CadenceFailures).Once()
+			},
+		},
+		"uncategorized error": {
+			input: errors.New("some random error"),
+			expectation: func(scope *mocks.Scope) {
+				scope.Mock.On("IncCounter", metrics.CadenceFailures).Once()
+			},
+		},
+	}
+
+	for name, td := range tests {
+		t.Run(name, func(t *testing.T) {
+			scope := mocks.Scope{}
+			td.expectation(&scope)
+			h := handlerImpl{
+				Resource: resource.NewTest(gomock.NewController(t), 0),
+			}
+			h.error(td.input, &scope, "some-domain", "some-wf", "some-run")
```

Comment:
I don't love the mocking library either, I'd like to get rid of it in favour of our normal gomock one, but it's what's already here, so I didn't try and change it. 

```
    --- FAIL: TestCorrectUseOfErrorHandling/uncategorized_error (0.00s)
panic: 
assert: mock: I don't know what to return because the method call was unexpected.
	Either do Mock.On("IncCounter").Return(...) first, or remove the IncCounter() call.
	This method was unexpected:
		IncCounter(int)
		0: 1
	panic: 
assert: mock: I don't know what to return because the method call was unexpected.
	Either do Mock.On("IncCounter").Return(...) first, or remove the IncCounter() call.
	This method was unexpected:
		IncCounter(int)
		0: 1
```

You're right there's nothing ensuring there's a counter call, let me add an assertion on that.

Reply:
Added `scope.Mock.AssertCalled` below to ensure that completely ommiting a call to counter is not missed